File: examples/checks/complex_numbers.py
# Independent Python checks for the complex_numbers example.
from __future__ import annotations

import cmath
import math
import re

from .common import run_checks


def to_complex(pair: dict) -> complex:
    return complex(float(pair["Re"]), float(pair["Im"]))


def parse_answer_values(answer: str) -> dict[str, complex]:
    values = {}
    pattern = r"^(C[0-9]+).*?=\s+([-0-9.]+)\s+([+-])\s+([0-9.]+)i\s*$"
    for ident, real, sign, imag in re.findall(pattern, answer, flags=re.MULTILINE):
        values[ident] = complex(float(real), float(imag) * (1 if sign == "+" else -1))
    return values


def close(a: complex, b: complex, tol: float = 5e-12) -> bool:
    return abs(a.real - b.real) <= tol and abs(a.imag - b.imag) <= tol


def run(ctx):
    data = ctx.load_input()
    expected = {}
    recomputed = {}
    polar_angles = {}
    for item in data["Exponents"]:
        ident = item["ID"]
        base = to_complex(item["Base"])
        power = to_complex(item["Power"])
        expected[ident] = to_complex(item["Expected"])
        recomputed[ident] = base ** power
        polar_angles[ident] = cmath.phase(base)
    for item in data["Inverses"]:
        ident = item["ID"]
        z = to_complex(item["Input"])
        expected[ident] = to_complex(item["Expected"])
        recomputed[ident] = cmath.asin(z) if item["Operation"] == "asin" else cmath.acos(z)

    reported = parse_answer_values(ctx.answer)
    asin_value = recomputed["C5"]
    acos_value = recomputed["C6"]

    checks = [
        ("principal polar angles for -1, e, and i match the N3 dial cases", abs(polar_angles["C1"] - math.pi) <= 1e-12 and abs(polar_angles["C2"]) <= 1e-12 and abs(polar_angles["C3"] - math.pi / 2) <= 1e-12),
        ("all four complex exponentiation answers match independent complex arithmetic", all(close(recomputed[item["ID"]], expected[item["ID"]]) for item in data["Exponents"])),
        ("i^i and e^(-pi/2) recompute to the same real value", abs(recomputed["C3"] - recomputed["C4"]) <= 1e-12 and abs(recomputed["C3"].imag) <= 1e-12),
        ("asin(2) and acos(2) match independent inverse-trig recomputation", close(recomputed["C5"], expected["C5"]) and close(recomputed["C6"], expected["C6"])),
        ("sin(asin(2)) and cos(acos(2)) round-trip to 2+0i", abs(cmath.sin(asin_value) - 2) <= 1e-12 and abs(cmath.cos(acos_value) - 2) <= 1e-12),
        ("asin(2) + acos(2) equals pi/2 with cancelling imaginary parts", abs((asin_value + acos_value).real - math.pi / 2) <= 1e-12 and abs((asin_value + acos_value).imag) <= 1e-12),
        ("all six reported complex outputs match recomputation to displayed precision", set(reported) == set(recomputed) and all(abs(reported[k] - recomputed[k]) <= 5e-12 for k in reported)),
        ("the report contains four exponentiation and two inverse-trig queries", len(data["Exponents"]) == 4 and len(data["Inverses"]) == 2 and "primitive test facts : 6" in ctx.reason),
        ("all recomputed outputs are finite real/imaginary pairs", all(math.isfinite(z.real) and math.isfinite(z.imag) for z in recomputed.values())),
    ]
    return run_checks(checks)

File: examples/checks/complex_numbers.go
package main

import (
	"math"
	"math/cmplx"
)

func checkComplexNumbers(ctx *Context) []Check {
	d := ctx.M()
	expected := map[string]complex128{}
	recomputed := map[string]complex128{}
	angles := map[string]float64{}
	for _, it := range maps(d["Exponents"]) {
		id := str(it["ID"])
		base := toComplex(asMap(it["Base"]))
		pow := toComplex(asMap(it["Power"]))
		expected[id] = toComplex(asMap(it["Expected"]))
		recomputed[id] = cmplx.Pow(base, pow)
		angles[id] = cmplx.Phase(base)
	}
	for _, it := range maps(d["Inverses"]) {
		id := str(it["ID"])
		z := toComplex(asMap(it["Input"]))
		expected[id] = toComplex(asMap(it["Expected"]))
		if str(it["Operation"]) == "asin" {
			recomputed[id] = cmplx.Asin(z)
		} else {
			recomputed[id] = cmplx.Acos(z)
		}
	}
	reported := parseComplexAnswer(ctx.Answer)
	asinv := recomputed["C5"]
	acosv := recomputed["C6"]
	expOK := true
	for _, it := range maps(d["Exponents"]) {
		id := str(it["ID"])
		expOK = expOK && complexClose(recomputed[id], expected[id], 5e-12)
	}
	reportOK := len(reported) == len(recomputed)
	for k, v := range reported {
		reportOK = reportOK && complexClose(v, recomputed[k], 5e-12)
	}
	finite := true
	for _, z := range recomputed {
		finite = finite && !math.IsNaN(real(z)) && !math.IsNaN(imag(z)) && !math.IsInf(real(z), 0) && !math.IsInf(imag(z), 0)
	}
	return []Check{{"principal polar angles for -1, e, and i match the N3 dial cases", math.Abs(angles["C1"]-math.Pi) <= 1e-12 && math.Abs(angles["C2"]) <= 1e-12 && math.Abs(angles["C3"]-math.Pi/2) <= 1e-12}, {"all four complex exponentiation answers match independent complex arithmetic", expOK}, {"i^i and e^(-pi/2) recompute to the same real value", cmplx.Abs(recomputed["C3"]-recomputed["C4"]) <= 1e-12 && math.Abs(imag(recomputed["C3"])) <= 1e-12}, {"asin(2) and acos(2) match independent inverse-trig recomputation", complexClose(recomputed["C5"], expected["C5"], 5e-12) && complexClose(recomputed["C6"], expected["C6"], 5e-12)}, {"sin(asin(2)) and cos(acos(2)) round-trip to 2+0i", cmplx.Abs(cmplx.Sin(asinv)-2) <= 1e-12 && cmplx.Abs(cmplx.Cos(acosv)-2) <= 1e-12}, {"asin(2) + acos(2) equals pi/2 with cancelling imaginary parts", math.Abs(real(asinv+acosv)-math.Pi/2) <= 1e-12 && math.Abs(imag(asinv+acosv)) <= 1e-12}, {"all six reported complex outputs match recomputation to displayed precision", reportOK}, {"the report contains four exponentiation and two inverse-trig queries", len(maps(d["Exponents"])) == 4 && len(maps(d["Inverses"])) == 2 && contains(ctx.Reason, "primitive test facts : 6")}, {"all recomputed outputs are finite real/imaginary pairs", finite}}
}
